src/source/channel_router.rs

```diff
@@ -0,0 +1,306 @@
+            current_channel: channel_count,
+            // this will cause the input buffer to fill on first call to next()
+            channel_count,
+            // channel_count is redundant, it's implicit in the channel_map dimensions
+            // but maybe it's saving us some time, we do check this value a lot.
+            input_buffer: vec![],
+            receiver: rx,
+        };
+
+        (controller, source)
+    }
+
+    /// Destroys this router and returns the underlying source.
+    #[inline]
+    pub fn into_inner(self) -> I {
+        self.input
+    }
+
+    /// Get mutable access to the inner source.
+    #[inline]
+    pub fn inner_mut(&mut self) -> &mut I {
+        &mut self.input
+    }
+}
+
+impl<I> Source for ChannelRouterSource<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    #[inline]
+    fn current_frame_len(&self) -> Option<usize> {
+        self.input.current_frame_len()
+    }
+
+    #[inline]
+    fn channels(&self) -> u16 {
+        self.channel_count
+    }
+
+    #[inline]
+    fn sample_rate(&self) -> u32 {
+        self.input.sample_rate()
+    }
+
+    #[inline]
+    fn total_duration(&self) -> Option<std::time::Duration> {
+        self.input.total_duration()
+    }
+}
+
+impl<I> Iterator for ChannelRouterSource<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current_channel >= self.channel_count {
+            // We've reached the end of the frame, time to grab another one from the input
+            let input_channels = self.input.channels() as usize;
+
+            // This might be too fussy, a source should never break a frame in the middle of an
+            // audio frame.
+            let samples_to_take = min(
+                input_channels,
+                self.input.current_frame_len().unwrap_or(usize::MAX),
+            );
+
+            // fill the input buffer. If the input is exhausted and returning None this will make
+            // the input buffer zero length
+            self.input_buffer = self.inner_mut().take(samples_to_take).collect();
+
+            self.current_channel = 0;
+
+            for change in self.receiver.try_iter() {
+                self.channel_map[change.0][change.1] = change.2;
+            }
+        }
+
+        // Find the output sample for current_channel
+        let retval = self
+            .input_buffer
+            .iter()
+            .zip(&self.channel_map)
+            .map(|(in_sample, input_gains)| {
+                // the way this works, the input_buffer need not be totally full, the router will
+                // work with whatever samples are available and the missing samples will be assumed
+                // to be equilibrium.
+                let gain = input_gains[self.current_channel as usize];
+                in_sample.amplify(gain)
+            })
+            .reduce(|a, b| a.saturating_add(b));
```

Comment:
Since this is real-time I'd prefer this to be more explicit. Maybe the matrix should be transposed to streamline this. This way instead of zip it would select a gain vector and then do the dot product of coefficients by input_buffer (it can be even optimized with SIMD).

Reply:
I do not want to overburden @iluvcapra, this is already quite a big effort! So we can always leave optimizations for later and merge this first. Someone else can work on them too (maybe @PetrGlad). Let me know what you prefer @iluvcapra.